src/metabase/query_processor/streaming/xlsx.clj

```diff
@@ -447,38 +478,143 @@
     (.setAutoFilter ^SXSSFSheet sheet (new CellRangeAddress 0 0 0 (dec col-count)))
     (.createFreezePane ^SXSSFSheet sheet 0 1)))
 
+(defn- cell-range
+  [rows]
+  (let [x (dec (count (first rows)))
+        y (dec (count rows))]
+    (CellRangeAddress.
+         0 ;; first row
+         y ;; last row
+         0 ;; first col
+         x ;; last col
+         )))
+
+(defn- cell-range->area-ref
+  [cell-range]
+  (AreaReference. (.formatAsString ^CellRangeAddress cell-range) SpreadsheetVersion/EXCEL2007))
+
+;; Possible Functions: https://poi.apache.org/apidocs/dev/org/apache/poi/ss/usermodel/DataConsolidateFunction.html
+;; I'm only including the keys that seem to work for our Pivot Tables as of 2024-06-06
+(def ^:private agg-fn-k->pivot-agg-fn-enum
+  {:sum    DataConsolidateFunction/SUM
+   :avg    DataConsolidateFunction/AVERAGE
+   :min    DataConsolidateFunction/MIN
+   :max    DataConsolidateFunction/MAX
+   :count  DataConsolidateFunction/COUNT
+   :stddev DataConsolidateFunction/STD_DEV})
+
+(defn- native-pivot
+  [rows
+   {:keys [pivot-grouping-key] :as pivot-spec}
+   {:keys [ordered-cols col-settings viz-settings]}]
+  (let [idx-shift                       (fn [indices]
+                                          (map (fn [idx]
+                                                 (if (> idx pivot-grouping-key)
+                                                   (dec idx)
+                                                   idx)) indices))
+        ordered-cols                    (vec (m/remove-nth pivot-grouping-key ordered-cols))
+        {:keys [pivot-rows
+                pivot-cols
+                pivot-measures
+                aggregation-functions]} (-> pivot-spec
+                                            (update :pivot-rows idx-shift)
+                                            (update :pivot-cols idx-shift)
+                                            (update :pivot-measures idx-shift)
+                                            (update :aggregation-functions #(vec (m/remove-nth pivot-grouping-key %))))
```

Comment:
the destructuring feels a bit unneccesary here, what do you think of doing

```suggestion
       pivot-rows (idx-shift (:pivot-rows pivot-spec))
       ...
```

instead?